### ISSUE TITLE
Building from source with commit 1d7caa58e fails on macOS 13.3.1 (a) and Perl 5.30

### ISSUE THREAD

**Issue by al3xandru** [vim/vim#12758]
### Steps to reproduce

1. checkout commit 1d7caa58e (tag: v9.0.1681) patch 9.0.1681: Build Failure with Perl 5.38 
2. `make clean`
3. `make install`

### Expected behaviour

Builds successfully

### Version of Vim

1d7caa58e

### Environment

macOS 13.3.1 (a)

### Logs and stack traces

```shell
gcc -c -I. -Ilibvterm/include -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1        -DINLINE="" -DVSNPRINTF=vim_vsnprintf -DSNPRINTF=vim_snprintf -DIS_COMBINING_FUNCTION=utf_iscomposing_uint -DWCWIDTH_FUNCTION=utf_uint2cells -o objects/vterm_vterm.o libvterm/src/vterm.c
/usr/bin/perl -e 'unless ( $] >= 5.005 ) { for (qw(na defgv errgv)) { print "#define PL_$_ $_\n" }}' > auto/if_perl.c
/usr/bin/perl /System/Library/Perl/5.30/ExtUtils/xsubpp -prototypes -typemap \
            /System/Library/Perl/5.30/ExtUtils/typemap if_perl.xs >> auto/if_perl.c
gcc -c -I. -DDYNAMIC_PERL_DLL=\"libperl.dylib\"   -g -DPERL_USE_SAFE_PUTENV  -iwithsysroot /System/Library/Perl/5.30/darwin-thread-multi-2level/CORE  -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1         -o objects/if_perl.o auto/if_perl.c
if_perl.xs:666:1: error: redefinition of 'S_SvREFCNT_dec'
S_SvREFCNT_dec(pTHX_ SV *sv)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/inline.h:209:1: note: previous definition is here
S_SvREFCNT_dec(pTHX_ SV *sv)
^
if_perl.xs:686:1: error: redefinition of 'S_TOPMARK'
S_TOPMARK(pTHX)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/inline.h:156:1: note: previous definition is here
S_TOPMARK(pTHX)
^
if_perl.xs:696:1: error: redefinition of 'S_POPMARK'
S_POPMARK(pTHX)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/inline.h:166:1: note: previous definition is here
S_POPMARK(pTHX)
^
3 errors generated.
make[1]: *** [objects/if_perl.o] Error 1
make: *** [install] Error 2
…/vim ❯
```

**Comment by trehousecomicsDev** [comment]
The developer has unfortunately passed away :(

**Comment by kohnish** [comment]
The following line looks suspicious to me 
But maybe mention your configure option and xcode version.  

 ```
-#ifdef DYNAMIC_PERL
+#if (PERL_REVISION == 5) && (PERL_VERSION >= 18)  
```

**Comment by chrisbra** [comment]
ping @pheiduck

**Comment by pheiduck** [comment]
> The following line looks suspicious to me But maybe mention your configure option and xcode version.
> 
> ```
> -#ifdef DYNAMIC_PERL
> +#if (PERL_REVISION == 5) && (PERL_VERSION >= 18)  
> ```

@k-takata is working on a Patch: https://github.com/vim/vim/pull/12755 which revert this change.
I guess it was a mistake during debug by me.

**Comment by pheiduck** [comment]
Anyway, I can not reproduce it:
macOS 13.5 + Perl 5.30 + latest vim commit
```
make clean
./configure --enable-perlinterp
make
```
Result no build errors.

**Comment by al3xandru** [comment]
My Makefile has the following options:

```
CONFIGURE_OPTS = --with-features=huge --enable-multibyte --enable-terminal --enable-cscope 
CONFIGURE_OPTS += --with-compiledby="$(VERGIT)@$(VERDATE)"
CONFIGURE_OPTS += --enable-pythoninterp=dynamic 
CONFIGURE_OPTS += --enable-python3interp=dynamic   
CONFIGURE_OPTS += --enable-rubyinterp=dynamic  
CONFIGURE_OPTS += --enable-perlinterp=dynamic
ifeq ($(ARCH),x86_64)
	CONFIGURE_OPTS += --prefix=/usr/local --enable-luainterp=dynamic --with-lua-prefix=/usr/local
	export vi_cv_path_python3=$(HOME)/.pyenv/versions/$(VIM_BUILD_PY3VER)/bin/python3
	export vi_cv_dll_name_python3=$(HOME)/.pyenv/versions/$(VIM_BUILD_PY3VER)/lib/libpython$(VIM_BUILD_PY3LIB).dylib
	export vi_cv_path_python3_conf=$(HOME)/.pyenv/versions/$(VIM_BUILD_PY3VER)/lib/python$(VIM_BUILD_PY3LIB)/config-$(VIM_BUILD_PY3LIB)-darwin
endif
ifeq ($(ARCH),arm64)
	CONFIGURE_OPTS += --prefix=/opt/homebrew --enable-luainterp=dynamic --with-lua-prefix=/opt/homebrew
	export PATH := "/opt/homebrew/opt/python@$(VIM_BUILD_PY3LIB)/bin:/System/Library/Framework/Python.framework/Versions/$(VIM_BUILD_PY2LIB)/bin":$(PATH)
endif
```

The build that failed was built without calling `./configure`.

I will try now with the HEAD.

**Comment by al3xandru** [comment]
The build fails with `--enable-perlinterp=dynamic` and builds successfully with `--enable-perlinterp`. Which confirms the suspicion a commenter posted above.

**Comment by pheiduck** [comment]
> The build fails with `--enable-perlinterp=dynamic` and builds successfully with `--enable-perlinterp`. Which confirms the suspicion a commenter posted above.

Yep, it fails:
`./configure --enable-gui=no --with-features=Huge --enable-perlinterp=dynamic`
If I revert to:
`#ifdef DYNAMIC_PERL`
it's working again.

**Comment by chrisbra** [comment]
Hm,
not sure. I tried this patch:
```patch
diff --git a/src/if_perl.xs b/src/if_perl.xs
index 6c1003c41..9e1754231 100644
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -40,7 +40,7 @@
 /* Work around for perl-5.18.
  * Don't include "perl\lib\CORE\inline.h" for now,
  * include it after Perl_sv_free2 is defined. */
-#if (PERL_REVISION == 5) && (PERL_VERSION >= 18)
+#ifdef DYNAMIC_PERL
 # define PERL_NO_INLINE_FUNCTIONS
 #endif

```

and it still fails with:

```
gcc -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2  -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1        -o objects/pathdef.o auto/pathdef.c
gcc -c -I. -DDYNAMIC_PERL_DLL=\"libperl.a\"  -fwrapv  -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64   -I/home/chrisbra/local/stow/perl532/lib/5.32.1/x86_64-linux/CORE  -Iproto -DHAVE_CONFIG_H     -g -O2  -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1         -o objects/if_perl.o auto/if_perl.c
In file included from /home/chrisbra/local/stow/perl532/lib/5.32.1/x86_64-linux/CORE/perl.h:3920,
                 from if_perl.xs:62:
if_perl.xs: In function ‘XS_VIM_Buffers’:
/home/chrisbra/local/stow/perl532/lib/5.32.1/x86_64-linux/CORE/op.h:88:18: warning: implicit declaration of function ‘Perl_gimme_V’ [-Wimplicit-function-declaration]
   88 | #define GIMME_V  Perl_gimme_V(aTHX)
      |                  ^~~~~~~~~~~~
/home/chrisbra/local/stow/perl532/lib/5.32.1/x86_64-linux/CORE/op.h:88:18: note: in definition of macro ‘GIMME_V’
   88 | #define GIMME_V  Perl_gimme_V(aTHX)
      |                  ^~~~~~~~~~~~
gcc -c -I. -Iproto -DHAVE_CONFIG_H     -g -O2  -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1        version.c -o objects/version.o
link.sh: $LINK_AS_NEEDED set to 'yes': invoking linker directly.
  gcc   -Wl,-E   -L/usr/local/lib -Wl,--as-needed       -o vim objects/alloc.o objects/arabic.o objects/arglist.o objects/autocmd.o objects/beval.o objects/buffer.o objects/change.o objects/blob.o objects/blowfish.o objects/cindent.o objects/clientserver.o objects/clipboard.o objects/cmdexpand.o objects/cmdhist.o objects/crypt.o objects/crypt_zip.o objects/debugger.o objects/dict.o objects/diff.o objects/digraph.o objects/drawline.o objects/drawscreen.o objects/edit.o objects/eval.o objects/evalbuffer.o objects/evalfunc.o objects/evalvars.o objects/evalwindow.o objects/ex_cmds.o objects/ex_cmds2.o objects/ex_docmd.o objects/ex_eval.o objects/ex_getln.o objects/fileio.o objects/filepath.o objects/findfile.o objects/float.o objects/fold.o objects/getchar.o objects/gui_xim.o objects/hardcopy.o objects/hashtab.o objects/help.o objects/highlight.o objects/if_cscope.o objects/if_xcmdsrv.o objects/indent.o objects/insexpand.o objects/list.o objects/locale.o objects/logfile.o objects/map.o objects/mark.o objects/match.o objects/mbyte.o objects/memline.o objects/menu.o objects/misc1.o objects/misc2.o objects/mouse.o objects/move.o objects/normal.o objects/ops.o objects/option.o objects/optionstr.o objects/os_unix.o objects/pathdef.o objects/popupmenu.o objects/popupwin.o objects/profiler.o objects/pty.o objects/quickfix.o objects/regexp.o objects/register.o objects/screen.o objects/scriptfile.o objects/search.o objects/session.o objects/sha256.o objects/sign.o objects/sound.o objects/spell.o objects/spellfile.o objects/spellsuggest.o objects/strings.o objects/syntax.o objects/tag.o objects/term.o objects/terminal.o objects/testing.o objects/textformat.o objects/textobject.o objects/textprop.o objects/time.o objects/typval.o objects/ui.o objects/undo.o objects/usercmd.o objects/userfunc.o objects/version.o objects/vim9class.o objects/vim9cmds.o objects/vim9compile.o objects/vim9execute.o objects/vim9expr.o objects/vim9instr.o objects/vim9script.o objects/vim9type.o objects/viminfo.o objects/window.o objects/bufwrite.o  objects/vterm_encoding.o objects/vterm_keyboard.o objects/vterm_mouse.o objects/vterm_parser.o objects/vterm_pen.o objects/vterm_screen.o objects/vterm_state.o objects/vterm_unicode.o objects/vterm_vterm.o   objects/if_perl.o objects/if_perlsfio.o      objects/netbeans.o objects/job.o objects/channel.o objects/xdiffi.o objects/xemit.o objects/xprepare.o objects/xutils.o objects/xhistogram.o objects/xpatience.o  objects/charset.o objects/json.o objects/main.o objects/memfile.o objects/message.o    -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo  -lselinux -lcanberra -lsodium -lrt -lacl -lattr -lgpm -ldl   -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/home/chrisbra/local/stow/perl532/lib/5.32.1/x86_64-linux/CORE -lperl -ldl -lm -lcrypt
/usr/bin/ld: objects/if_perl.o: in function `XS_VIM_Windows':
/home/chrisbra/code/vim-upstream/src/if_perl.xs:1703: undefined reference to `Perl_gimme_V'
/usr/bin/ld: objects/if_perl.o: in function `XS_VIM_Buffers':
/home/chrisbra/code/vim-upstream/src/if_perl.xs:1652: undefined reference to `Perl_gimme_V'
collect2: error: ld returned 1 exit status
link.sh: Linking failed
make: *** [Makefile:2062: vim] Error 1
```

using perl 5.32

**Comment by chrisbra** [comment]
This seems to work for me with Perl 5.32 and Perl 5.37:

```patch
diff --git a/src/if_perl.xs b/src/if_perl.xs
index 6c1003c41..dedfde683 100644
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -40,7 +40,7 @@
 /* Work around for perl-5.18.
  * Don't include "perl\lib\CORE\inline.h" for now,
  * include it after Perl_sv_free2 is defined. */
-#if (PERL_REVISION == 5) && (PERL_VERSION >= 18)
+#ifdef DYNAMIC_PERL
 # define PERL_NO_INLINE_FUNCTIONS
 #endif

@@ -709,6 +709,10 @@ S_POPMARK(pTHX)
 #  define Perl_POPMARK S_POPMARK
 # endif

+# if (PERL_REVISION == 5) && (PERL_VERSION >= 37)
+#  define GIMME GIMME_V
+# endif
+
 /* perl-5.34 needs Perl_SvTRUE_common; used in SvTRUE_nomg_NN */
 # if (PERL_REVISION == 5) && (PERL_VERSION == 34)
 PERL_STATIC_INLINE bool
@@ -1649,7 +1653,7 @@ Buffers(...)
     PPCODE:
     if (items == 0)
     {
-       if (GIMME_V == G_SCALAR)
+       if (GIMME == G_SCALAR)
        {
            i = 0;
            FOR_ALL_BUFFERS(vimbuf)
@@ -1700,7 +1704,7 @@ Windows(...)
     PPCODE:
     if (items == 0)
     {
-       if (GIMME_V == G_SCALAR)
+       if (GIMME == G_SCALAR)
            XPUSHs(sv_2mortal(newSViv(win_count())));
        else
        {
```

But I don't really know the perl internals.

**Comment by pheiduck** [comment]
@chrisbra that Patch LGTM :)

**Comment by chrisbra** [comment]
should be fixed by #12755

**Comment by ychin** [comment]
I mentioned in this comment (https://github.com/vim/vim/pull/12755#issuecomment-1689829921) but this isn't working properly for Perl 5.30 and only working by accident (since it's actually linking statically instead of dynamically due to the linker flags). See that comment for more info. I can work on a fix.
